hip/components/cooperative_groups.hip.hpp

```diff
@@ -0,0 +1,442 @@
+template <size_type Size>
+struct thread_block_tile
+    : detail::enable_extended_shuffle<detail::thread_block_tile<Size>> {
+    using detail::enable_extended_shuffle<
+        detail::thread_block_tile<Size>>::enable_extended_shuffle;
+};
+
+
+// Only support tile_partition with 1, 2, 4, 8, 16, 32, 64 (hip).
+template <size_type Size, typename Group>
+__device__ __forceinline__ gko::xstd::enable_if_t<
+    (Size <= kernels::hip::hip_config::warp_size) &&
+        (kernels::hip::hip_config::warp_size % Size == 0),
+    thread_block_tile<Size>>
+tiled_partition(const Group &)
+{
+    return thread_block_tile<Size>();
+}
+
+
+namespace detail {
+
+
+template <size_type Size>
+struct is_group_impl<thread_block_tile<Size>> : std::true_type {};
+template <size_type Size>
+struct is_synchronizable_group_impl<thread_block_tile<Size>> : std::true_type {
+};
+template <size_type Size>
+struct is_communicator_group_impl<thread_block_tile<Size>> : std::true_type {};
+
+
+}  // namespace detail
+
+
+class thread_block {
+    friend __device__ __forceinline__ thread_block this_thread_block();
+
+public:
+    __device__ __forceinline__ unsigned thread_rank() const noexcept
+    {
+        return data_.rank;
+    }
+
+    __device__ __forceinline__ unsigned size() const noexcept
+    {
+        return data_.size;
+    }
+
+    __device__ __forceinline__ void sync() const noexcept { __syncthreads(); }
+
+private:
+    __device__ thread_block()
+        : data_{static_cast<unsigned>(blockDim.x * blockDim.y * blockDim.z),
+                static_cast<unsigned>(
+                    threadIdx.x +
+                    blockDim.x * (threadIdx.y + blockDim.y * threadIdx.z))}
+    {}
+    struct alignas(8) {
+        unsigned size;
+        unsigned rank;
+    } data_;
+};
+
+
+__device__ __forceinline__ thread_block this_thread_block()
+{
+    return thread_block();
+}
+
+
+namespace detail {
+
+template <>
+struct is_group_impl<thread_block> : std::true_type {};
+template <>
+struct is_synchronizable_group_impl<thread_block> : std::true_type {};
+
+
+}  // namespace detail
+
+
+/**
+ * This is a limited implementation of the CUDA grid_group that works even on
+ * devices that do not support device-wide synchronization and without special
+ * kernel launch syntax.
+ *
+ * Note that this implementation (as well as the one from CUDA's cooperative
+ * groups) does not support large grids, since it uses 32 bits to represent
+ * sizes and ranks, while at least 73 bits (63 bit grid + 10 bit block) would
+ * have to be used to represent the full space of thread ranks.
+ */
+class grid_group {
+    friend __device__ grid_group this_grid();
+
+public:
+    __device__ unsigned size() const noexcept { return data_.size; }
+
+    __device__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+private:
+    __device__ grid_group()
+        : data_{blockDim.x * blockDim.y * blockDim.z * gridDim.x * gridDim.y *
+                    gridDim.z,
```

Comment:
This looks really weird. Maybe you can format it better and wrap it in `// clang-format off` and `// clang-format on` comments ?